### DISCUSSION TITLE
Add DistributionFeatureMatching in QuaPy

### DISCUSSION THREAD

**Discussion by BastienDussap** [HLT-ISTI/QuaPy#20] [Ideas]
Hi everyone,

As I have discussed with some of you at ECML/PKDD, I would like to add **DistributionFeatureMatching** (from _Label Shift Quantification with Robustness Guarantees via Distribution Feature Matching_: https://arxiv.org/pdf/2306.04376.pdf).

The general method uses an embedding $\Phi$ that takes a distribution and returns the mean $\Phi(\mathbb{P}) = \mathbb{E}_{\mathbb{P}}[\Phi(X)]$.
Using this we solve:
$\underset{\alpha \in \Delta^c}{\arg \min}\ \left|  \sum \alpha_i \Phi(\hat{\mathbb{P}}_i) - \Phi(\hat{\mathbb{Q}}) \right|^2$
where $\Phi(\hat{\mathbb{P}}_i)$ is the embedding of the class $i$ in the source (training set) and $\Phi(\hat{\mathbb{Q}})$ is the embedding of the target (testing set).

Here is what I want to implement:

- **DFM**, where the user gives in _init_ the function $\Phi$ (the vectorization function)
- **RFFM** (RandomFourierFeaturesMatching) where I use $\Phi(x) = \sqrt{\frac{2}{D}} [\cos(w_i^Tx), \sin(w_i^Tx)]_{i=1}^{D/2}$, for $D$ a given even number and $w_i$ sampled from a Gaussian distribution. This choice of function comes from the theory of the _RandomFourierFeatures_ (https://proceedings.neurips.cc/paper/2007/file/013a006f03dbc5392effeb8f18fda755-Paper.pdf) that allows to do kernel methods in linear time (see appendix C in https://arxiv.org/pdf/2306.04376.pdf).
- **KernelQuantifier**, where the function $\Phi$ is the KernelMeanEmbedding (see https://www.nowpublishers.com/article/Details/MAL-060) assosiated to a kernel $k$. In that case, we can't compute $\Phi(\mathbb{P})$ but we can compute the scalar product $\langle\Phi(\mathbb{P}), \Phi(\mathbb{Q})\rangle$, which is enough to solve the minimisation problem. I want to implement the Gaussian kernel (in that case we recover https://proceedings.mlr.press/v32/iyer14.html) and the Energy Kernel (in that case we recover https://search.ieice.org/bin/summary.php?id=e99-d_1_176) and a version where you give your own kernel function.

Note: to solve the minimisation problem, we must rely on a QP solver, in my case I used **cvxopt**(https://cvxopt.org/)

I've already coded (and partially tested) these methods and made them compatible with _Quapy.LabelledCollection_.

What do you think about this ?
How should we proceed to add these methods, do I create a new branch **DFM** with these methods ? (I'm not familiar with project management using git)

Sincerely
Bastien.

**Comment by AlexMoreo** [comment]
Hi Bastien!

Thanks for contacting us. Sure, your contributions to QuaPy are more than welcome. 
Before getting started, though, I would like to share with you some basic guidelines.
I will be happy to assist you with the git-related issues, don't worry about that.
I will come back to you ASAP.

Congrats on the well-deserved award at ECML-PKDD!
Best regards,
Alejandro

**Comment by mirkobunse** [comment]
Hi @BastienDussap , Hi @AlexMoreo ,

our team has integrated the kernel mean embedding and the random Fourier features into the [qunfold](https://github.com/mirkobunse/qunfold) Python package.

Since our package de-couples the minimization of a loss from the representation of the data, we only had to implement the kernel and RFF representations you propose in your paper, following the implementations you already had published. The cool thing about having these methods in qunfold is that they can now be easily re-composed with other loss functions and pre-processors. All quantifiers in qunfold can readily be used in QuaPy.

We are also planning to provide general support for the generalized label shift scenario that your paper discusses.

**Comment by AlexMoreo** [comment]
Amazing!

Thanks for sharing it! We should start integrating stuff into quapy 
(which seems very easy thanks to your wrapper), and taking advantage of 
the cross-fertilization with qunfold.

Cheers,
Alex

El 2024-01-10 14:44, Mirko Bunse escribió:
> Hi @BastienDussap [1] , Hi @AlexMoreo [2] ,
> 
> our team has integrated the kernel mean embedding and the random
> Fourier features into the qunfold [3] Python package.
> 
> Since our package de-couples the minimization of a loss from the
> representation of the data, we only had to implement the kernel and
> RFF representations you propose in your paper, following the
> implementations you already had published. The cool thing about having
> these methods in qunfold is that they can now be easily re-composed
> with other loss functions and pre-processors. All quantifiers in
> qunfold can readily be used in QuaPy.
> 
> We are also planning to provide general support for the generalized
> label shift scenario that your paper discusses.
> 
> --
> Reply to this email directly, view it on GitHub [4], or unsubscribe
> [5].
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
> 
> Links:
> ------
> [1] https://github.com/BastienDussap
> [2] https://github.com/AlexMoreo
> [3] https://github.com/mirkobunse/qunfold
> [4] 
> https://github.com/HLT-ISTI/QuaPy/discussions/20#discussioncomment-8080233
> [5] 
> https://github.com/notifications/unsubscribe-auth/AELURNBQ7MGMEYMXQDQHZPLYN2LMFAVCNFSM6AAAAAA5GI3MPGVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4DAOBQGIZTG

-- 
Alejandro Moreo
Istituto di Scienza e Tecnologie dell'Informazione (ISTI)
Area della Ricerca CNR di Pisa - via G. Moruzzi 1, 56124 Pisa Italy
